### PR TITLE
Build docs in container

### PR DESCRIPTION
**Pull request by maximiliankolb**
This PR adds a `Containerfile` to build a container image & minimal usage instructions on how to build Foreman documentation without installing any other applications apart from docker/ a docker alternative. If someone is kind enough to rerun both commands on Fedora, I would **love** to replace `docker` with `podman`.

I hope this makes it easier for everyone to try and build their branch/contribution locally first before creating a PR.

On a sidenote: We could also create a separate repo (like `foreman-documentation-container-image`) to massively speed up Github actions. It'll be faster and save resources.

### CODE REVIEW

File: guides/README.md [hunk 1]
```diff
@@ -67,6 +67,17 @@ It's also possible to check links; the following command will check all links ex
...
+## Building Locally Using a Container
+
+Build container image:
+
+    docker build --no-cache --progress=plain --file Containerfile --tag foreman_documentation:1.0 .
+
+Build Foreman documentation.
+Run this command in your `foreman-documentation` repository:
+
+    rm -rf guides/build && docker run --rm -v $(pwd):/foreman-documentation foreman_documentation:1.0 make html
```
Comment (ekohl) [inline guides/README.md:79]: On SELinux enabled systems (at least with podman) you need to pass `-Z` for correct labels. On non-SELinux you can't pass it. Sadly that's the way it is.

File: Containerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FROM fedora:34
+
+RUN dnf upgrade -y && \
+    dnf install -y findutils-1:4.8.0-2.fc34 \
+                   linkchecker-9.4.0-12.20191005.d13b3f5.fc34 \
+                   make-1:4.3-5.fc34 \
+                   ruby-3.0.2-149.fc34
+
+RUN gem install --pre asciidoctor:2.0.16 \
+                      asciidoctor-pdf:1.6.0
```
[the remark this answers is not in the document]
Comment (maximiliankolb) [reply Containerfile:10]: I pinned packages to exact versions to allow others to rebuild to same "exact" container image. Removed.

I've went this way because of the instructions in `guides/README.md`. I've added a commit to correct the pages names accordingly.

File: guides/README.md [hunk 3]
```diff
@@ -67,6 +67,17 @@ It's also possible to check links; the following command will check all links ex
 
 	make linkchecker
 
+## Building Locally Using a Container
+
+Build container image:
+
+    docker build --no-cache --progress=plain --file Containerfile --tag foreman_documentation:1.0 .
```
[the remark this answers is not in the document]
Comment (maximiliankolb) [reply guides/README.md:74]: IMHO `Containerfile` is a more application agnostic term. See also http://docs.podman.io/en/latest/markdown/podman-build.1.html

File: guides/README.md (hunk 1 above)
Comment (maximiliankolb) [reply guides/README.md:79]: I don't know what this means & I cannot reproduce this on my machine. I've added this as a comment, but feel free to correct/reword this once more.

**Review by melcorr — COMMENTED** [review]
I can't get this to work... but then that could just be me. 
Would you mind if we wait for @lzap to come back off leave before merging this? 
He might have some ideas.

**Comment by maximiliankolb** [comment]
> I can't get this to work... but then that could just be me.
> Would you mind if we wait for @lzap to come back off leave before merging this?
> He might have some ideas.

Do you want to share error messages or commands you've tried so far? Maybe my instructions are incomplete.

**Comment by melcorr** [comment]
> > I can't get this to work... but then that could just be me.
> > Would you mind if we wait for @lzap to come back off leave before merging this?
> > He might have some ideas.
> 
> Do you want to share error messages or commands you've tried so far? Maybe my instructions are incomplete.

I will try this on a VM tomorrow to try and rule out my own dodgy setup.

File: guides/README.md [hunk 4]
```diff
@@ -67,6 +67,19 @@ It's also possible to check links; the following command will check all links ex
 
 	make linkchecker
 
+## Building Locally Using a Container
+
```
Comment (melcorr) [inline guides/README.md:71]: Can we cater to the lowest common denominator because some writers might not be as technical as you, and just add that ensure that you have Docker installed.

**Review by melcorr — COMMENTED** [review]
I have gotten this to work! 
I just had conflicting files from an older version of docker. 
Cleaned it all and installed the latest and this worked.
I would still appreciate an ack from @lzap just incase I am missing something. 
However, if he doesn't get to it, I will merge soon as this is great and won't harm anything. 
One tiny suggestion.

File: guides/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (melcorr) [reply guides/README.md:74]: On Fedora, I had to run this with `sudo`

File: guides/README.md (hunk 4 above)
Comment (maximiliankolb) [reply guides/README.md:71]: I've added minimal usage instructions. Afterall, using a container should make things easier.

Comment (melcorr) [reply guides/README.md:71]: Thanks. I get you completely. And am totally satisfied with the minimal instructions. I think that we are covered by this.

File: guides/README.md [hunk 5]
```diff
@@ -67,6 +67,22 @@ It's also possible to check links; the following command will check all links ex
...
+You can build Foreman documentation locally using a container image.
+This requires the cloned git repository plus an application such as Podman or Docker to build and run container images.
+
+1. Build container image:
+
+       docker build --no-cache --progress=plain --file Containerfile --tag foreman_documentation:1.0 .
```
[the remark this answers is not in the document]
Comment (melcorr) [reply guides/README.md:77]: My first run had errors because of `--progress` also, I remember.

File: guides/README.md [hunk 6]
```diff
@@ -67,6 +67,22 @@ It's also possible to check links; the following command will check all links ex
...
+
+2. Build Foreman documentation.
+   Run this command in your `foreman-documentation` repository:
+
+       rm -rf guides/build && docker run --rm -v $(pwd):/foreman-documentation foreman_documentation:1.0 make html
+
+   Add the `-Z` flag on SELinux enabled systems.
```
Comment (lzap) [inline guides/README.md:84]: Is this related?

```
[lzap@nuc foreman-documentation]$ docker run --rm -v $(pwd):/foreman-documentation foreman_documentation:1.0 make html
make: stat: Makefile: Permission denied
make: *** No rule to make target 'html'.  Stop.
```

Comment (melcorr) [reply guides/README.md:84]: I can only do this with sudo.

File: guides/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (maximiliankolb) [reply guides/README.md:77]: Fixed.

File: guides/README.md (hunk 6 above)
Comment (maximiliankolb) [reply guides/README.md]: I cannot reproduce this, but I've added a note to run this as root on systems using SELinux. I am unsure if this makes it user friendly.

File: guides/README.md (hunk 1 above)
Comment (lzap) [reply guides/README.md]: I am getting this, is it related?

```
$ docker run --rm -v $(pwd):/foreman-documentation foreman_documentation:1.0 make html
make: stat: Makefile: Permission denied
make: *** No rule to make target 'html'.  Stop.
```

Option -Z is not known to my podman. Where exactly should I pass this?

File: guides/README.md (hunk 6 above)
Comment (lzap) [reply guides/README.md]: Ok I know what is going on. Users do not map correctly, my ID is 1001 (yeah I am special) and this does not map to root:

```
$ docker run --rm -v $(pwd):/foreman-documentation -it foreman_documentation:1.0 /bin/bash
[root@21ecf14c6fda /]# ls -la /
total 16
dr-xr-xr-x.   1 root root  42 Aug 12 16:25 .
dr-xr-xr-x.   1 root root  42 Aug 12 16:25 ..
drwxrwxr-x.   1 1001 1001 224 Aug 12 14:25 foreman-documentation
...
```

Within the container environment, the effective user (which is UID=0) cannot do anything with the contents. I am not container expert, not sure what to do here.

File: guides/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximiliankolb) [reply guides/README.md]: Added `sudo`.

**Comment by lzap** [comment]
What is the status? I cannot get this working on my Fedora in rootless mode. Requiring root is something I would like to avoid - it got me into troubles already as build/ directory was created as root and subsequent builds failed because permission errors.

File: guides/README.md (hunk 6 above)
Comment (ekohl) [reply guides/README.md]: We have a similar script in pulpcore-packaging and here's how we deal with SELinux:
https://github.com/theforeman/pulpcore-packaging/blob/2b69c4af86d45833d3530ba9e78db301fc25abe5/automation/update_packaging.sh#L16-L22

**Comment by maximiliankolb** [comment]
I ran this on macOS and it works for me. Unfortunately, I am unable to run this on other systems like Fedora or Ubuntu. I try my best to make it work for everyone, regardless of their system environment (which is the whole point of containers IMHO).

I would be happy if someone made **suggestions** on this PR.

**Comment by melcorr** [comment]
I think that is a fair point @maximiliankolb 
I have no background with containers, so can't really report more than what happens when I test it. 
I used a VM in libvirt to test on Fedora. A VM is always an option for further testing. 
I will try and find someone who is better versed than we seem to be. Sorry and I understand the frustration with this.

File: guides/README.md [hunk 7]
```diff
@@ -67,6 +67,22 @@ It's also possible to check links; the following command will check all links ex
...
 	make linkchecker
 
+## Building Locally Using a Container
+
+You can build Foreman documentation locally using a container image.
+This requires the cloned git repository plus an application such as Podman or Docker to build and run container images.
+
+1. Build container image:
+
+       sudo podman build --no-cache --file Containerfile --tag foreman_documentation:1.0 .
```
Comment (ekohl) [inline guides/README.md:77]: Why `no-cache`?

On Fedora 34, this also works for me:
```suggestion
       podman build --tag foreman_documentation:1.0 .
```

In fact, the that this build is with sudo and the one below is without is guaranteed to give problems. You must run either both as root or both as a normal user.

File: guides/README.md [hunk 8]
```diff
@@ -67,6 +67,22 @@ It's also possible to check links; the following command will check all links ex
...
+
+2. Build Foreman documentation.
+   Run this command in your `foreman-documentation` repository:
+
+       rm -rf guides/build && podman run --rm -v $(pwd):/foreman-documentation foreman_documentation:1.0 make html
+
+   On SELinux enabled systems, run this command as root using `sudo` and add the `-Z` flag.
```
Comment (ekohl) [inline guides/README.md:84]:
```suggestion
   On SELinux enabled systems, run this command:
   
        rm -rf guides/build && podman run --rm -v $(pwd):/foreman-documentation:Z foreman_documentation:1.0 make html 
```

**Review by ekohl — COMMENTED** [review]
I think nowadays on Fedora rootless podman should work out of the box.

File: guides/README.md (hunk 7 above)
Comment (maximiliankolb) [reply guides/README.md:77]: Fixed.

File: guides/README.md (hunk 8 above)
Comment (maximiliankolb) [reply guides/README.md:84]: Fixed.

**Comment by maximiliankolb** [comment]
Rebased to master.

File: guides/README.md (hunk 1 above)
Comment (maximiliankolb) [reply guides/README.md]: I've added separate usage instructions for SELinux enabled systems thanks to @ekohl

**Comment by xprazak2** [comment]
Works for me on Fedora without sudo, @lzap are you still getting premission issues?
docs

**Review by ekohl — APPROVED** [review]

**Comment by melcorr** [comment]
Thanks to everyone who helped with this and for your patience @maximiliankolb !
